core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp

```diff
@@ -751,24 +751,61 @@ struct ParallelReduceSpecialize<FunctorType, TeamPolicyInternal<PolicyArgs...>,
                            ? OpenMPTargetExec::MAX_ACTIVE_TEAMS
                            : league_size;
 
+    const size_t shmem_size_L0 = p.scratch_size(0, team_size);
+    const size_t shmem_size_L1 = p.scratch_size(1, team_size);
     OpenMPTargetExec::resize_scratch(PolicyType::member_type::TEAM_REDUCE_SIZE,
-                                     0, 0);
+                                     shmem_size_L0, shmem_size_L1);
     void* scratch_ptr = OpenMPTargetExec::get_scratch_ptr();
 
     ValueType result = ValueType();
 
+    // Maximum active teams possible.
+    int max_active_teams = OpenMPTargetExec::MAX_ACTIVE_THREADS / team_size;
+
+    int* lock_array = OpenMPTargetExec::get_lock_array(max_active_teams);
+
+#pragma omp target teams distribute parallel for is_device_ptr(lock_array)
+    for (int i = 0; i < max_active_teams; ++i) {
+      lock_array[i] = 0;
+    }
+
 #pragma omp target teams distribute num_teams(nteams) thread_limit(team_size) \
          map(to:f) map(tofrom:result) reduction(+: result) \
-    is_device_ptr(scratch_ptr)
+    is_device_ptr(scratch_ptr, lock_array)
     for (int i = 0; i < league_size; i++) {
       ValueType inner_result = ValueType();
+      int shmem_block_index = -1, lock_team = 99999, iter = -1;
+
+      // Loop as long as a shmem_block_index is not found.
+      while (shmem_block_index == -1) {
+        // Loop until a lock can be acquired on a team
+        while (lock_team != 1) {
+          // Avoid tripping over other team's block index.
+          // Find a free block index.
+          iter = (omp_get_team_num() % max_active_teams);
```

Comment:
shouldn't this happen outside the while loop? Is that also wrong in the ParallelFor?

Reply:
also why do we need two while loops. Isn't the only way to escape the inner while loop for lock_team==1 in which case shmem_block_indx != -1?

Reply:
Yes the additional loop is not needed. Deleted it.